File: rd-net/Test.RdFramework/Contexts/RdPerContextMapTest.cs
using System.Collections.Generic;
using JetBrains.Rd;
using JetBrains.Rd.Impl;
using NUnit.Framework;

namespace Test.RdFramework.Contexts
{
  public class RdPerContextMapTest : RdFrameworkTestBase
  {
    public override void SetUp()
    {
      base.SetUp();
      ServerWire.AutoTransmitMode = true;
      ClientWire.AutoTransmitMode = true;
    }

    [Test]
    public void TestOnStructMap()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var client1Cid = "Client-1";
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      ServerProtocol.Contexts.GetValueSet(key).Add(server1Cid);

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);

      serverMap[server1Cid][1] = "test";
      
      Assert.True(clientMap.TryGetValue(server1Cid, out var map) && map != null);
      Assert.AreEqual("test", clientMap[server1Cid][1]);

      ClientProtocol.Contexts.GetValueSet(key).Add(client1Cid);
      
      Assert.True(serverMap.TryGetValue(client1Cid, out var map1) && map1 != null);
      Assert.False(serverMap[client1Cid].ContainsKey(1));
    }

    [Test]
    public void TestLateBind01()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var client1Cid = "Client-1";
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      ServerProtocol.Contexts.GetValueSet(key).Add(server1Cid);
      
      serverMap[server1Cid][1] = "test";
      ClientProtocol.Contexts.GetValueSet(key).Add(client1Cid);
      
      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);
      
      Assert.True(clientMap.TryGetValue(server1Cid, out var map) && map != null);
      Assert.AreEqual("test", clientMap[server1Cid][1]);

      Assert.True(serverMap.TryGetValue(client1Cid, out var map1) && map1 != null);
      Assert.False(serverMap[client1Cid].ContainsKey(1));
    }
    
    [Test]
    public void TestLateBind02()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var client1Cid = "Client-1";
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      // no protocol value set value - pre-bind value will be lost
      
      serverMap[server1Cid][1] = "test";

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);
      
      Assert.False(clientMap.TryGetValue(server1Cid, out _));

      Assert.False(serverMap.TryGetValue(client1Cid, out _));
    }
    
    [Test]
    public void TestLateBind03()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var client1Cid = "Client-1";
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      ServerProtocol.Contexts.GetValueSet(key).Add(server1Cid);
      ServerProtocol.Contexts.GetValueSet(key).Add(client1Cid);

      var log = new List<string>();
      
      serverMap.View(LifetimeDefinition.Lifetime, (lifetime, s, _) =>
      {
        log.Add("Add " + s);
        lifetime.OnTermination(() => log.Add("Remove " + s));
      });

      serverMap[server1Cid][1] = "test";
      serverMap[client1Cid][1] = "test";

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);
      
      Assert.AreEqual(new []{"Add " + server1Cid, "Add " + client1Cid}, log);
    }
    
    [Test]
    public void TestLateBind04()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var client1Cid = "Client-1";
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      ServerProtocol.Contexts.GetValueSet(key).Add(server1Cid);

      var log = new List<string>();
      
      serverMap.View(LifetimeDefinition.Lifetime, (lifetime, s, _) =>
      {
        log.Add("Add " + s);
        lifetime.OnTermination(() => log.Add("Remove " + s));
      });

      serverMap[server1Cid][1] = "test";
      serverMap[client1Cid][1] = "test";

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);
      
      Assert.AreEqual(new []{"Add " + server1Cid, "Add " + client1Cid, "Remove " + client1Cid}, log);
    }
    
    [Test]
    public void TestLateBind05()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";

      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);
      
      var log = new List<string>();
      
      serverMap.View(LifetimeDefinition.Lifetime, (lifetime, s, _) =>
      {
        log.Add("Add " + s);
        lifetime.OnTermination(() => log.Add("Remove " + s));
      });
      
      ServerProtocol.Contexts.GetValueSet(key).Add(server1Cid);

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);
      
      serverMap[server1Cid][1] = "test";
      
      Assert.AreEqual(new []{"Add " + server1Cid}, log);
    }
    
    [Test]
    public void TestLateBind06()
    {
      var key = RdContextBasicTest.TestKeyHeavy.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";

      var log = new List<string>();
      
      serverMap.View(LifetimeDefinition.Lifetime, (lifetime, s, _) =>
      {
        log.Add("Add " + s);
        lifetime.OnTermination(() => log.Add("Remove " + s));
      });
      
      key.RegisterOn(ClientProtocol.Serializers);
      ServerProtocol.Contexts.RegisterContext(key);
      ClientProtocol.Contexts.RegisterContext(key);

      BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
      BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);

      using (key.UpdateValue(server1Cid))
      {
        ServerProtocol.Wire.Send(ServerProtocol.Identities.Mix(RdId.Nil, 10), _ => { });
      }
      
      Assert.True(ServerProtocol.Contexts.GetValueSet(key).Contains(server1Cid));
      
      Assert.AreEqual(new []{"Add " + server1Cid}, log);
    }
    
    [Test]
    public void TestValueSetChangesInContext()
    {
      var key1 = RdContextBasicTest.TestKeyHeavy.Instance;
      var key2 = RdContextBasicTest.TestKey2.Instance;

      var serverMap = new RdPerContextMap<string, RdMap<int, string>>(key1, _ => NewRdMap<int, string>());
      var clientMap = new RdPerContextMap<string, RdMap<int, string>>(key1, _ => NewRdMap<int, string>());

      var server1Cid = "Server-1";
      var server2Cid = "Server-2";
      var server3Cid = "Server-3";
      var server4Cid = "Server-4";

      var log = new List<string>();
      
      serverMap.View(LifetimeDefinition.Lifetime, (lifetime, s, _) =>
      {
        log.Add("Add " + s);
        lifetime.OnTermination(() => log.Add("Remove " + s));
      });

      using (key1.UpdateValue(server1Cid))
      using (key2.UpdateValue(server1Cid))
      {
        key1.RegisterOn(ClientProtocol.Serializers);
        key2.RegisterOn(ClientProtocol.Serializers);
        ServerProtocol.Contexts.RegisterContext(key1);
        ServerProtocol.Contexts.RegisterContext(key2);
        ClientProtocol.Contexts.RegisterContext(key1);

        BindToClient(LifetimeDefinition.Lifetime, clientMap, 1);
        BindToServer(LifetimeDefinition.Lifetime, serverMap, 1);

        ServerProtocol.Contexts.GetValueSet(key1).Add(server2Cid);
        using (key1.UpdateValue(server4Cid))
        {
          ServerProtocol.Contexts.GetValueSet(key1).Add(server3Cid);
        }
      }

      Assert.False(ServerProtocol.Contexts.GetValueSet(key1).Contains(server1Cid));
      Assert.False(ServerProtocol.Contexts.GetValueSet(key2).Contains(server1Cid));
      
      Assert.AreEqual(new []{"Add " + server2Cid, "Add " + server3Cid}, log);
    }
  }
}

File: rd-kt/rd-framework/src/test/kotlin/com/jetbrains/rd/framework/test/cases/contexts/RdPerContextMapTest.kt
@file:Suppress("UNCHECKED_CAST")
package com.jetbrains.rd.framework.test.cases.contexts

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.bindTopLevel
import com.jetbrains.rd.framework.base.static
import com.jetbrains.rd.framework.impl.RdMap
import com.jetbrains.rd.framework.impl.RdPerContextMap
import com.jetbrains.rd.framework.test.util.DynamicEntity
import com.jetbrains.rd.framework.test.util.RdFrameworkTestBase
import com.jetbrains.rd.util.assert
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test


class RdPerContextMapTest : RdFrameworkTestBase() {
    private val serializers = Serializers()

    override fun createSerializers(isServer: Boolean): ISerializers {
        return serializers
    }

    @Test
    fun testOnStructMap() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, FrameworkMarshallers.String).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, FrameworkMarshallers.String).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val client1Cid = "Client-1"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        serverProtocol.contexts.getValueSet(key).add(server1Cid)

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")
        serverMap[server1Cid]!![1] = "test"

        assert(clientMap[server1Cid]!![1] == "test")

        clientProtocol.contexts.getValueSet(key).add(client1Cid)

        assert(serverMap[client1Cid] != null)
        assert(serverMap[client1Cid]!![1] == null)
    }

    @Test
    fun testOnDynamicMap() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val client1Cid = "Client-1"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        serverProtocol.contexts.getValueSet(key).add(server1Cid)

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")
        serverMap[server1Cid]!![1] = DynamicEntity("test")

        assert(clientMap[server1Cid]!![1]!!.foo.value == "test")

        clientProtocol.contexts.getValueSet(key).add(client1Cid)

        assert(serverMap[client1Cid] != null)
        assert(serverMap[client1Cid]!![1] == null)
    }

    @Test
    fun testLateBind01() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val client1Cid = "Client-1"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        serverProtocol.contexts.getValueSet(key).add(server1Cid)

        serverMap[server1Cid]!![1] = DynamicEntity("test")
        clientProtocol.contexts.getValueSet(key).add(client1Cid)

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        assert(clientMap[server1Cid]!![1]!!.foo.value == "test")

        assert(serverMap[client1Cid] != null)
        assert(serverMap[client1Cid]!![1] == null)
    }

    @Test
    fun testLateBind02() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        // no protocol value set value - pre-bind value will be lost

        serverMap[server1Cid]!![1] = DynamicEntity("test")

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        assert(serverMap[server1Cid] == null)
        assert(clientMap[server1Cid] == null)
    }

    @Test
    fun testLateBind03() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val server2Cid = "Server-2"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        serverProtocol.contexts.getValueSet(key).addAll(setOf(server1Cid, server2Cid))

        val log = ArrayList<String>()

        serverMap.view(serverLifetime) { entryLt, k, _ ->
            log.add("Add $k")
            entryLt.onTermination {
                log.add("Remove $k")
            }
        }

        serverMap[server1Cid]!![1] = DynamicEntity("test")
        serverMap[server2Cid]!![1] = DynamicEntity("test")

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        assertEquals(listOf("Add $server1Cid", "Add $server2Cid"), log)
    }

    @Test
    fun testLateBind04() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val server2Cid = "Server-2"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        serverProtocol.contexts.getValueSet(key).addAll(setOf(server1Cid))

        val log = ArrayList<String>()

        serverMap.view(serverLifetime) { entryLt, k, _ ->
            log.add("Add $k")
            entryLt.onTermination {
                log.add("Remove $k")
            }
        }

        serverMap[server1Cid]!![1] = DynamicEntity("test")
        serverMap[server2Cid]!![1] = DynamicEntity("test")

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        assertEquals(listOf("Add $server1Cid", "Add $server2Cid", "Remove $server2Cid"), log)
    }

    @Test
    fun testLateBind05() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        val log = ArrayList<String>()

        serverMap.view(serverLifetime) { entryLt, k, _ ->
            log.add("Add $k")
            entryLt.onTermination {
                log.add("Remove $k")
            }
        }

        serverProtocol.contexts.getValueSet(key).addAll(setOf(server1Cid))

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        serverMap[server1Cid]!![1] = DynamicEntity("test")

        assertEquals(listOf("Add $server1Cid"), log)
    }

    @Test
    fun testLateBind06() {
        val key = ContextsTest.TestKeyHeavy

        val serverMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"

        val log = ArrayList<String>()

        serverMap.view(serverLifetime) { entryLt, k, _ ->
            log.add("Add $k")
            entryLt.onTermination {
                log.add("Remove $k")
            }
        }

        serverProtocol.contexts.registerContext(key)
        clientProtocol.contexts.registerContext(key)

        clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
        serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

        key.updateValue(server1Cid).use {
            serverProtocol.wire.send(serverProtocol.identity.mix(RdId.Null, 10)) {} // trigger key addition by protocol write
        }

        assertTrue(serverProtocol.contexts.getValueSet(key).contains(server1Cid))

        serverMap[server1Cid]!![1] = DynamicEntity("test")

        assertEquals(listOf("Add $server1Cid"), log)
    }

    @Test
    fun testValueSetChangesInContext() {
        val key1 = object : ThreadLocalRdContext<String>("test-key1", true, FrameworkMarshallers.String) {}
        val key2 = object : ThreadLocalRdContext<String>("test-key2", true, FrameworkMarshallers.String) {}

        val serverMap = RdPerContextMap(key1) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1)
        val clientMap = RdPerContextMap(key1) { RdMap(FrameworkMarshallers.Int, DynamicEntity as ISerializer<DynamicEntity<String>>).apply { master = it } }.static(1).apply { master = false }

        val server1Cid = "Server-1"
        val server2Cid = "Server-2"
        val server3Cid = "Server-3"
        val server4Cid = "Server-4"

        val log = ArrayList<String>()

        serverMap.view(serverLifetime) { entryLt, k, _ ->
            log.add("Add $k")
            entryLt.onTermination {
                log.add("Remove $k")
            }
        }

        key2.updateValue(server1Cid).use {
            key1.updateValue(server1Cid).use {

                serverProtocol.contexts.registerContext(key1)
                serverProtocol.contexts.registerContext(key2)
                clientProtocol.contexts.registerContext(key1)

                clientMap.bindTopLevel(clientLifetime, clientProtocol, "map")
                serverMap.bindTopLevel(serverLifetime, serverProtocol, "map")

                serverProtocol.contexts.getValueSet(key1).add(server2Cid)
            }

            key1.updateValue(server4Cid).use {
                serverProtocol.contexts.getValueSet(key1).add(server3Cid)
            }
        }

        assertFalse(serverProtocol.contexts.getValueSet(key1).contains(server1Cid))
        assertFalse(serverProtocol.contexts.getValueSet(key2).contains(server1Cid))
        assertEquals(listOf("Add $server2Cid", "Add $server3Cid"), log)
    }
}
